cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return val.ID
+	}
+	return nil
+}
+
+// GetLoadBalancerState will get the state for the load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerState(name string) *infrav1beta2.VPCLoadBalancerState {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		return &val.State
+	}
+	return nil
+}
+
+// GetLoadBalancerHost will return the hostname of load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerHost(name string) *string {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		return val.Hostname
+	}
+	return nil
+}
+
+// SetCOSBucketStatus set the COS bucket id.
+func (s *PowerVSClusterScope) SetCOSBucketStatus(cosInstanceID string, controllerCreated bool) {
+	if s.IBMPowerVSCluster.Status.COSInstance == nil {
+		s.IBMPowerVSCluster.Status.COSInstance = &infrav1beta2.ResourceReference{
+			ID:                &cosInstanceID,
+			ControllerCreated: &controllerCreated,
+		}
+	} else {
+		// set the controllerCreated only if its not set
+		if !*s.IBMPowerVSCluster.Status.COSInstance.ControllerCreated {
+			s.IBMPowerVSCluster.Status.COSInstance.ControllerCreated = &controllerCreated
+		}
+		s.IBMPowerVSCluster.Status.COSInstance.ID = &cosInstanceID
+	}
+}
+
+// GetResourceGroupID returns the resource group id.
+func (s *PowerVSClusterScope) GetResourceGroupID() (string, error) {
+	if s.IBMPowerVSCluster.Spec.ResourceGroup != nil && s.IBMPowerVSCluster.Spec.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Spec.ResourceGroup.ID, nil
+	}
+	if s.IBMPowerVSCluster.Status.ResourceGroup != nil && s.IBMPowerVSCluster.Status.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Status.ServiceInstance.ID, nil
+	}
+	resourceGroupID, err := s.fetchResourceGroupID()
+	if err != nil {
+		return "", err
+	}
+	s.SetStatus(ResourceGroup, infrav1beta2.ResourceReference{ID: &resourceGroupID, ControllerCreated: pointer.Bool(false)})
+	return resourceGroupID, nil
+}
+
+// ReconcileServiceInstance reconciles service instance.
+func (s *PowerVSClusterScope) ReconcileServiceInstance() error {
+	serviceInstanceID := s.GetServiceInstanceID()
+	if serviceInstanceID != "" {
+		s.Info("Service instance id is set", "id", serviceInstanceID)
+		serviceInstance, _, err := s.ResourceClient.GetResourceInstance(&resourcecontrollerv2.GetResourceInstanceOptions{
+			ID: &serviceInstanceID,
+		})
+		if err != nil {
+			return err
+		}
+		if serviceInstance == nil {
+			return fmt.Errorf("error failed to get service instance with id %s", serviceInstanceID)
+		}
+		if *serviceInstance.State != string(infrav1beta2.ServiceInstanceStateActive) {
+			return fmt.Errorf("service instance not in active state, current state: %s", *serviceInstance.State)
+		}
+		s.Info("Found service instance and its in active state", "id", serviceInstanceID)
```

Comment:
I see this can be a code you can run post creation of resource in the end as well.

Reply:
applies this comment for all other reconciles as well